### ISSUE TITLE
HL Iterator Performance

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#7587]
The following Haxe code triggers many allocations in HL:

```haxe

class Main {
	
	static inline function find<T>( a : Iterable<T>, p : T -> Bool ) : T {
		var found : T = cast null;
		for( x in a )
			if( p(x) ) {
				found = x;
				break;
			}
		return found;
	}
	
	static function main() {
		var a = [1,2,0,0,0,0,0,0,0,0,0,0,0,3];
		hl.Profile.enable = true;
				
		var v = find(a, x -> x > 2);
	
		hl.Profile.dump();
		
	}
	
}
```

The main problem is that we don't allow methods in HL virtuals, leading to Iterator being alias of Dynamic, so calls to `next()` / `hasNext()` allocate a closure for each access.

**Comment by ncannasse** [comment]
This is fixed by allow Var/Method in TAnon to be part of the virtual.
However this cause the following unit tests to crash:

```
src/unit/TestMain.hx:50: Generated at: 2018-10-31 15:41:49
src/unit/TestMain.hx:52: START
src/unit/Test.hx:276: ABORTED : Access violation in unit.TestEReg.test
src/unit/Test.hx:279: STACK :

Called from local function #1349 (unit/TestEReg.hx line 101)
Called from local function #1350 (D:\Projects\haxe\std/hl/_std/Date.hx line 0)
Called from EReg.map (D:\Projects\haxe\std/hl/_std/EReg.hx line 173)
Called from unit.TestEReg.test (unit/TestEReg.hx line 103)
src/unit/Test.hx:276: ABORTED : Access violation in unit.TestMatch.testExtractors
src/unit/Test.hx:279: STACK :

Called from local function #3484 (unit/TestMatch.hx line 479)
Called from unit.TestMatch.testExtractors (unit/TestMatch.hx line 487)
src/unit/Test.hx:276: ABORTED : Access violation in unit.TestSpecification.testArray_unit_hx
src/unit/Test.hx:279: STACK :

Called from local function #3590 (unitstd/Array.unit.hx line 260)
Called from local function #3591 (D:\Projects\haxe\std/hl/_std/Date.hx line 0)
Called from hl.types.ArrayObj.map (D:\Projects\haxe\std/hl/types/ArrayObj.hx line 231)
Called from unit.TestSpecification.testArray_unit_hx (unitstd/Array.unit.hx line 261)
src/unit/Test.hx:276: ABORTED : Access violation in unit.TestSpecification.testReflect_unit_hx
src/unit/Test.hx:279: STACK :

Called from unit.C2.set_propAcc (unit/TestSpecification.hx line 45)
Called from unit.$C2.__constructor__ (unit/TestSpecification.hx line 37)
Called from unit.TestSpecification.testReflect_unit_hx (unitstd/Reflect.unit.hx line 11)
src/unit/Test.hx:276: ABORTED : Access violation in unit.issues.Issue3113.test
src/unit/Test.hx:279: STACK :

Called from unit.issues.Issue3113.test (unit/issues/Issue3113.hx line 8)
src/unit/Test.hx:276: ABORTED : Access violation in unit.issues.Issue3967.testNicolas
src/unit/Test.hx:279: STACK :

Called from unit.issues.Issue3967.testNicolas (unit/issues/Issue3967.hx line 23)
src/unit/Test.hx:276: ABORTED : Access violation in unit.issues.Issue4910.test
src/unit/Test.hx:279: STACK :

Called from unit.issues.Issue4889.test (unit/issues/Issue4889.hx line 5)
Called from unit.issues.Issue4910.test (unit/issues/Issue4910.hx line 16)
src/unit/Test.hx:276: ABORTED : Access violation in unit.issues.Issue6379.test
src/unit/Test.hx:279: STACK :

Called from unit.issues.Issue6379.g (unit/issues/Issue6379.hx line 9)
Called from unit.issues.Issue6379.test (unit/issues/Issue6379.hx line 5)
src/unit/Test.hx:222: DONE [8064 tests]
src/unit/Test.hx:223: SUCCESS: false
```

**Comment by ncannasse** [comment]
Fixed with HL 1.9
